### ISSUE TITLE
`unicorn/no-useless-spread` does not allow transform the type

### ISSUE THREAD

**Issue by sausir1** [oxc-project/oxc#24107]
# Description

Noticed this after migrating from `eslint` to `oxclint`, that returning a `Set` and wanting to transform it to array using the spread syntax is flagged by the `unicorn/no-useless-spread` as `redundant`

## What's wrong

URL to reproduction: https://playground.oxc.rs/?t=formatter&lintRules=unicorn%2Fno-useless-spread&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Atrue%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22ts%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%22config%22%3A%22%7B%5C%22categories%5C%22%3A%7B%5C%22correctness%5C%22%3A%5C%22off%5C%22%7D%2C%5C%22rules%5C%22%3A%7B%5C%22unicorn%2Fno-useless-spread%5C%22%3A%5C%22error%5C%22%7D%7D%22%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=class+Something+%7B%0A++constructor%28%29+%7B%7D%0A%0A++filter%28%29%3A+Set%3Cstring%3E+%7B%0A++++const+items+%3D+%5B%271%27%2C%272%27%2C%273%27%5D%0A%0A++++return+new+Set%28items%29%0A++%7D%0A%7D%0A%0Aconst+smt+%3D+new+Something%28%29%0A%0Aconst+result%3A+string%5B%5D+%3D+%5B...smt.filter%28%29%5D+%2F%2F+why+this+gets+marked+as+error%3F%0A%0Aconst+result2+%3D+Array.from%28smt.filter%28%29%29+%2F%2F+this+doesn%27t%0A%0Aconst+set1+%3D+new+Set%28%5B%27foo%27%5D%29%0A%0Aconst+result3+%3D+%5B...set1%5D+%2F%2F+this+doesn%27t+too%0A%0Aconst+set2+%3D+smt.filter%28%29%0A%0Aconst+result4+%3D+%5B...set2%5D+%2F%2F+now+it+passes.%0A

## Expected behavior

The transform to `array` shouldn't be marked by the `no-useless-spread` rule if the type of returned item is different.

**Comment by kapobajza** [comment]
Hello @sausir1. From what I can see, this is flagged in `oxc` because the rule isn't type aware. However, the upstream ESLint rule is, as you can see here:
1. [The `isKnownArrayMethodClone` calls the `isArray` helper](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/58670ffae9f94f06745ba193aec0bd97f3a0a875/rules/no-useless-spread.js#L165-L175)
2. [The `isArray` helper is constructed from the `createTypeCheckers` helper](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/utils/is-array.js#L55C1-L73C4)
3. And finally [`createTypeCheckers` returns functions that are used for type checking](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/utils/type-helpers.js#L721-L733)

Therefore:

```javascript
class Something {
  constructor() {}

  filter(): Set<string> {
    const items = ['1','2','3']

    return new Set(items)
  }
}

const smt = new Something()

// There's no way to know what the actual return type of the `smt.filter` call is. It assumes that it's an array
// Additionally, `oxc` practically mistakes this as the `Array.filter` method call.
const result: string[] = [...smt.filter()]
```

But there's a workaround for this: just rename the method to something other than `filter`. Maybe `filterSet` or similar, and `oxc` will stop flagging it.

@camc314 How should we proceed with this issue? Do we port it to `tsgolint` or keep it as it s?

**Comment by camc314** [comment]
Hi! Thanks for making this issue. Thanks kapobajza for the research!

We currently have no plans to move this into tsgolint and make it type aware.

I would recommend disabling this rule as needed or alternatiely using a disable comment for that line/file.

We may make other changes in future to make it type aware.

Thanks!
